File: extras/aethon/android/app/src/main/java/com/firstham/aethergui/AppUpdateManager.java
package com.firstham.aethergui;

import android.app.DownloadManager;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.text.TextUtils;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.work.ExistingPeriodicWorkPolicy;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.PeriodicWorkRequest;
import androidx.work.WorkManager;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

final class AppUpdateManager {
    private static final String CHANNEL_ID = "aether_app_updates";
    private static final int NOTIFICATION_ID = 2901;
    private static final ExecutorService EXECUTOR = Executors.newSingleThreadExecutor();
    private static final AtomicBoolean CHECKING = new AtomicBoolean();

    interface Listener { void onComplete(); void onError(Throwable error); }

    static void initialize(Context context) {
        Context app = context.getApplicationContext();
        createNotificationChannel(app);
        reconcileDownload(app);
        boolean enabled = app.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE).getBoolean(UpdateConfig.KEY_AUTO_DOWNLOAD, false);
        setAutomaticChecks(app, enabled);
    }

    static void setAutomaticChecks(Context context, boolean enabled) {
        Context app = context.getApplicationContext();
        WorkManager manager = WorkManager.getInstance(app);
        if (!enabled) {
            manager.cancelUniqueWork("aether-update-check");
            manager.cancelUniqueWork("aether-update-startup");
            return;
        }
        PeriodicWorkRequest periodic = new PeriodicWorkRequest.Builder(UpdateWorker.class, 12, TimeUnit.HOURS)
                .setConstraints(UpdateWorker.constraints()).build();
        manager.enqueueUniquePeriodicWork("aether-update-check", ExistingPeriodicWorkPolicy.KEEP, periodic);
        manager.enqueueUniqueWork("aether-update-startup", ExistingWorkPolicy.KEEP, new OneTimeWorkRequest.Builder(UpdateWorker.class).setConstraints(UpdateWorker.constraints()).build());
    }

    static void checkNow(Context context, Listener listener) {
        Context app = context.getApplicationContext();
        if (!CHECKING.compareAndSet(false, true)) {
            new android.os.Handler(android.os.Looper.getMainLooper()).post(listener::onComplete);
            return;
        }
        app.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE).edit().putString("status", "checking").apply();
        sendState(app);
        EXECUTOR.execute(() -> {
            try {
                checkBlocking(app);
                android.os.Handler main = new android.os.Handler(android.os.Looper.getMainLooper());
                main.post(listener::onComplete);
            } catch (Throwable error) {
                markFailed(app);
                android.os.Handler main = new android.os.Handler(android.os.Looper.getMainLooper());
                main.post(() -> listener.onError(error));
            } finally {
                CHECKING.set(false);
            }
        });
    }

    static synchronized void checkBlocking(Context context) throws IOException {
        JSONArray releases = getJsonArray(UpdateConfig.API_URL);
        JSONObject release = null;
        String tag = "";
        JSONObject apk = null;
        for (int i = 0; i < releases.length(); i++) {
            JSONObject candidate = releases.optJSONObject(i);
            if (candidate == null) continue;
            String candidateTag = candidate.optString("tag_name", "").replaceFirst("^v", "");
            if (candidateTag.isEmpty() || compareVersions(candidateTag, BuildConfig.VERSION_NAME) <= 0) continue;
            JSONObject candidateApk = findAsset(candidate.optJSONArray("assets"), String.format(Locale.US, UpdateConfig.RELEASE_ASSET, candidateTag));
            if (candidateApk != null) {
                release = candidate;
                tag = candidateTag;
                apk = candidateApk;
                break;
            }
        }
        if (release == null) {
            for (int i = 0; i < releases.length(); i++) {
                JSONObject candidate = releases.optJSONObject(i);
                if (candidate == null) continue;
                String candidateTag = candidate.optString("tag_name", "").replaceFirst("^v", "");
                JSONObject candidateApk = findAsset(candidate.optJSONArray("assets"), String.format(Locale.US, UpdateConfig.RELEASE_ASSET, candidateTag));
                if (candidateApk != null) { release = candidate; tag = candidateTag; apk = candidateApk; break; }
            }
        }
        if (release == null || apk == null || tag.isEmpty()) throw new IOException("No Android update package is available");
        String downloadUrl = apk.optString("browser_download_url", "");
        if (!downloadUrl.startsWith(UpdateConfig.RELEASE_DOWNLOAD_PREFIX)) throw new IOException("The update URL is not an official Aethon release");
        String checksum = apk.optString("digest", "").replaceFirst("^sha256:", "");
        if (checksum.isEmpty()) {
            JSONObject sums = findAsset(release.optJSONArray("assets"), UpdateConfig.CHECKSUM_ASSET);
            if (sums != null) {
                String sumsUrl = sums.optString("browser_download_url", "");
                if (!sumsUrl.startsWith(UpdateConfig.RELEASE_DOWNLOAD_PREFIX)) throw new IOException("The checksum URL is not an official Aethon release");
                checksum = checksumFromFile(getText(sumsUrl), apk.optString("name"));
            }
        }
        if (checksum.isEmpty()) throw new IOException("The Android APK has no checksum");
        UpdateInfo info = new UpdateInfo(tag, release.optString("body", ""), downloadUrl, checksum);
        android.content.SharedPreferences prefs = context.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE);
        prefs.edit().putString(UpdateConfig.KEY_LATEST_VERSION, info.version).putString(UpdateConfig.KEY_RELEASE_NOTES, info.notes)
                .putString(UpdateConfig.KEY_DOWNLOAD_URL, info.downloadUrl).putString(UpdateConfig.KEY_CHECKSUM, info.checksum).apply();
        if (compareVersions(info.version, BuildConfig.VERSION_NAME) > 0) {
            prefs.edit().putString("status", prefs.getBoolean(UpdateConfig.KEY_AUTO_DOWNLOAD, false) ? "downloading" : "available").apply();
            notifyAvailable(context, info.version, info.notes);
            sendState(context);
            if (prefs.getBoolean(UpdateConfig.KEY_AUTO_DOWNLOAD, false)) startDownload(context, false);
        } else {
            prefs.edit().putString("status", "up_to_date").apply();
            sendState(context);
        }
    }

    static void markFailed(Context context) {
        context.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE).edit().putString("status", "failed").apply();
        sendState(context);
    }

    static boolean startDownload(Context context, boolean wifiOnly) {
        Context app = context.getApplicationContext();
        android.content.SharedPreferences prefs = app.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE);
        String url = prefs.getString(UpdateConfig.KEY_DOWNLOAD_URL, "");
        String version = prefs.getString(UpdateConfig.KEY_LATEST_VERSION, "");
        if (url.isEmpty() || version.isEmpty() || compareVersions(version, BuildConfig.VERSION_NAME) <= 0) return false;
        try {
            DownloadManager manager = (DownloadManager) app.getSystemService(Context.DOWNLOAD_SERVICE);
            long oldId = prefs.getLong(UpdateConfig.KEY_DOWNLOAD_ID, -1);
            if (oldId != -1 && isActive(manager, oldId)) return true;
            File dir = app.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS);
            if (dir == null) throw new IOException("External update storage is unavailable");
            if (!dir.exists() && !dir.mkdirs()) throw new IOException("Could not create the update directory");
            File apk = new File(dir, String.format(Locale.US, UpdateConfig.RELEASE_ASSET, version));
            if (apk.exists() && !apk.delete()) throw new IOException("Could not replace the previous update");
            DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url)).setTitle(app.getString(R.string.new_update_title))
                    .setDescription(app.getString(R.string.update_downloading)).setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
                    .setAllowedOverRoaming(false).setDestinationUri(Uri.fromFile(apk));
            request.setAllowedNetworkTypes(wifiOnly ? DownloadManager.Request.NETWORK_WIFI : DownloadManager.Request.NETWORK_WIFI | DownloadManager.Request.NETWORK_MOBILE);
            long id = manager.enqueue(request);
            prefs.edit().putLong(UpdateConfig.KEY_DOWNLOAD_ID, id).putString(UpdateConfig.KEY_APK_PATH, apk.getAbsolutePath()).putString("status", "downloading").apply();
            sendState(app);
            return true;
        } catch (Throwable error) {
            prefs.edit().putString("status", "download_failed").apply();
            sendState(app);
            notifyFailure(app, R.string.update_download_failed);
            return false;
        }
    }

    private static void reconcileDownload(Context context) {
        android.content.SharedPreferences prefs = context.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE);
        long id = prefs.getLong(UpdateConfig.KEY_DOWNLOAD_ID, -1);
        if (id < 0) return;
        DownloadManager manager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
        try (android.database.Cursor cursor = manager.query(new DownloadManager.Query().setFilterById(id))) {
            if (cursor == null || !cursor.moveToFirst()) {
                prefs.edit().remove(UpdateConfig.KEY_DOWNLOAD_ID).putString("status", "download_failed").apply();
                return;
            }
            int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS));
            if (status == DownloadManager.STATUS_SUCCESSFUL) {
                Intent completed = new Intent(context, AppUpdateReceiver.class).setAction(DownloadManager.ACTION_DOWNLOAD_COMPLETE)
                        .putExtra(DownloadManager.EXTRA_DOWNLOAD_ID, id);
                context.sendBroadcast(completed);
            } else if (status == DownloadManager.STATUS_PENDING || status == DownloadManager.STATUS_RUNNING || status == DownloadManager.STATUS_PAUSED) {
                prefs.edit().putString("status", "downloading").apply();
            } else {
                prefs.edit().remove(UpdateConfig.KEY_DOWNLOAD_ID).putString("status", "download_failed").apply();
            }
        } catch (RuntimeException error) {
            prefs.edit().putString("status", "download_failed").apply();
        }
    }

    @SuppressLint("MissingPermission")
    static void notifyAvailable(Context context, String version, String notes) {
        createNotificationChannel(context);
        android.content.SharedPreferences prefs = context.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE);
        if (version.equals(prefs.getString(UpdateConfig.KEY_NOTIFIED_VERSION, ""))) return;
        if (!notificationsAllowed(context)) return;
        Intent action = new Intent(context, AppUpdateReceiver.class).setAction(UpdateConfig.ACTION_DOWNLOAD);
        PendingIntent pending = PendingIntent.getBroadcast(context, 2902, action, PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
        String body = context.getString(R.string.new_update_versions, BuildConfig.VERSION_NAME, version) + "\n" + (TextUtils.isEmpty(notes) ? context.getString(R.string.no_release_notes) : notes.trim());
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID).setSmallIcon(R.drawable.ic_aethon_mono)
                .setContentTitle(context.getString(R.string.new_update_title)).setContentText(context.getString(R.string.new_update_versions, BuildConfig.VERSION_NAME, version))
                .setStyle(new NotificationCompat.BigTextStyle().bigText(body)).setOnlyAlertOnce(true).setAutoCancel(true).addAction(0, context.getString(R.string.start_update), pending);
        NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, builder.build());
        prefs.edit().putString(UpdateConfig.KEY_NOTIFIED_VERSION, version).apply();
    }

    @SuppressLint("MissingPermission")
    static void notifyInstallReady(Context context) {
        createNotificationChannel(context);
        if (!notificationsAllowed(context)) return;
        Intent action = new Intent(context, AppUpdateReceiver.class).setAction(UpdateConfig.ACTION_INSTALL);
        PendingIntent pending = PendingIntent.getBroadcast(context, 2903, action, PendingIntent.FLAG_IMMUTABLE | PendingIntent.FLAG_UPDATE_CURRENT);
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID).setSmallIcon(R.drawable.ic_aethon_mono)
                .setContentTitle(context.getString(R.string.new_update_title)).setContentText(context.getString(R.string.update_ready_install))
                .setAutoCancel(true).setOnlyAlertOnce(true).addAction(0, context.getString(R.string.install_update), pending);
        NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, builder.build());
    }

    @SuppressLint("MissingPermission")
    static void notifyFailure(Context context, int messageId) {
        createNotificationChannel(context);
        if (!notificationsAllowed(context)) return;
        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_ID).setSmallIcon(R.drawable.ic_aethon_mono)
                .setContentTitle(context.getString(R.string.app_updates)).setContentText(context.getString(messageId)).setAutoCancel(true).setOnlyAlertOnce(true);
        NotificationManagerCompat.from(context).notify(NOTIFICATION_ID, builder.build());
    }

    static void sendState(Context context) {
        Intent intent = new Intent(UpdateConfig.ACTION_STATE).setPackage(context.getPackageName());
        context.sendBroadcast(intent, AetherVpnService.INTERNAL_PERMISSION);
    }

    static int compareVersions(String left, String right) {
        String[] a = left.replaceFirst("^v", "").split("\\.");
        String[] b = right.replaceFirst("^v", "").split("\\.");
        for (int i = 0; i < Math.max(a.length, b.length); i++) {
            int av = i < a.length ? number(a[i]) : 0;
            int bv = i < b.length ? number(b[i]) : 0;
            if (av != bv) return Integer.compare(av, bv);
        }
        return 0;
    }

    static String checksumFromFile(String content, String assetName) {
        if (content == null) return "";
        for (String line : content.split("\\r?\\n")) {
            String[] parts = line.trim().split("\\s+", 2);
            if (parts.length == 2 && parts[1].trim().equals(assetName)) return parts[0].trim().toLowerCase(Locale.US);
        }
        return "";
    }

    private static int number(String value) { try { return Integer.parseInt(value.replaceAll("[^0-9].*", "")); } catch (Exception ignored) { return 0; } }
    private static JSONObject findAsset(JSONArray assets, String name) {
        if (assets == null) return null;
        for (int i = 0; i < assets.length(); i++) { JSONObject item = assets.optJSONObject(i); if (item != null && name.equals(item.optString("name"))) return item; }
        return null;
    }
    private static JSONObject getJson(String url) throws IOException {
        try { return new JSONObject(getText(url)); }
        catch (org.json.JSONException error) { throw new IOException("Invalid update metadata", error); }
    }

    private static JSONArray getJsonArray(String url) throws IOException {
        try { return new JSONArray(getText(url)); }
        catch (org.json.JSONException error) { throw new IOException("Invalid update metadata", error); }
    }
    private static String getText(String url) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();
        connection.setConnectTimeout(15_000); connection.setReadTimeout(30_000); connection.setRequestProperty("User-Agent", "Aethon-Android"); connection.setRequestProperty("Accept", "application/vnd.github+json");
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8))) {
            StringBuilder result = new StringBuilder(); String line; while ((line = reader.readLine()) != null) result.append(line).append('\n'); return result.toString();
        } finally { connection.disconnect(); }
    }
    private static boolean isActive(DownloadManager manager, long id) { try { android.database.Cursor cursor = manager.query(new DownloadManager.Query().setFilterById(id)); if (cursor == null) return false; try { if (!cursor.moveToFirst()) return false; int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS)); return status == DownloadManager.STATUS_PENDING || status == DownloadManager.STATUS_RUNNING || status == DownloadManager.STATUS_PAUSED; } finally { cursor.close(); } } catch (Exception ignored) { return false; } }
    static int downloadProgress(Context context) {
        android.content.SharedPreferences prefs = context.getSharedPreferences(UpdateConfig.PREFS, Context.MODE_PRIVATE);
        long id = prefs.getLong(UpdateConfig.KEY_DOWNLOAD_ID, -1);
        if (id < 0) return -1;
        DownloadManager manager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
        try (android.database.Cursor cursor = manager.query(new DownloadManager.Query().setFilterById(id))) {
            if (cursor == null || !cursor.moveToFirst()) return -1;
            int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS));
            if (status == DownloadManager.STATUS_SUCCESSFUL) return 100;
            if (status != DownloadManager.STATUS_PENDING && status != DownloadManager.STATUS_RUNNING && status != DownloadManager.STATUS_PAUSED) return -1;
            long total = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
            long downloaded = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
            return total > 0 ? (int) Math.min(99, downloaded * 100 / total) : 0;
        } catch (Exception ignored) { return -1; }
    }
    private static boolean notificationsAllowed(Context context) { return Build.VERSION.SDK_INT < 33 || context.checkSelfPermission("android.permission.POST_NOTIFICATIONS") == android.content.pm.PackageManager.PERMISSION_GRANTED; }
    private static void createNotificationChannel(Context context) { NotificationChannel channel = new NotificationChannel(CHANNEL_ID, context.getString(R.string.update_notification_channel), NotificationManager.IMPORTANCE_DEFAULT); channel.setDescription(context.getString(R.string.update_notification_channel_summary)); context.getSystemService(NotificationManager.class).createNotificationChannel(channel); }

    static String sha256(File file) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256"); byte[] buffer = new byte[32 * 1024]; int count;
        try (FileInputStream input = new FileInputStream(file)) { while ((count = input.read(buffer)) != -1) digest.update(buffer, 0, count); }
        StringBuilder result = new StringBuilder(); for (byte value : digest.digest()) result.append(String.format(Locale.US, "%02x", value)); return result.toString();
    }

    private AppUpdateManager() { }
}

File: extras/uac-sni-spoofer-android/app/src/main/java/com/uacspoofer/mobile/update/AppUpdateManager.kt
package com.uacspoofer.mobile.update

import android.app.Activity
import android.app.DownloadManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.Settings
import com.uacspoofer.mobile.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import kotlin.math.max

internal data class AppRelease(
    val version: String,
    val tagName: String,
    val title: String,
    val notes: String,
    val apkName: String,
    val apkUrl: String,
    val releaseUrl: String,
)

internal sealed interface UpdateCheckResult {
    data object Current : UpdateCheckResult
    data class Available(val release: AppRelease) : UpdateCheckResult
}

internal sealed interface UpdateUiState {
    data object Idle : UpdateUiState
    data object Checking : UpdateUiState
    data object UpToDate : UpdateUiState
    data class Available(val release: AppRelease) : UpdateUiState
    data class Downloading(val release: AppRelease, val progress: Int?) : UpdateUiState
    data class Ready(val release: AppRelease, val message: String) : UpdateUiState
    data class Error(val message: String, val release: AppRelease? = null) : UpdateUiState
}

internal enum class InstallLaunchResult {
    INSTALLER_OPENED,
    PERMISSION_REQUESTED,
}

internal class AppUpdateManager(context: Context) {
    private val appContext = context.applicationContext
    private val preferences = appContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)

    suspend fun checkForUpdate(): UpdateCheckResult = withContext(Dispatchers.IO) {
        val connection = (URL(LATEST_RELEASE_API).openConnection() as HttpURLConnection).apply {
            connectTimeout = NETWORK_TIMEOUT_MS
            readTimeout = NETWORK_TIMEOUT_MS
            requestMethod = "GET"
            setRequestProperty("Accept", "application/vnd.github+json")
            setRequestProperty("User-Agent", "UAC-SNI-Spoofer-Android/${BuildConfig.VERSION_NAME}")
        }
        try {
            val status = connection.responseCode
            if (status != HttpURLConnection.HTTP_OK) {
                error("GitHub Releases returned HTTP $status")
            }
            val release = parseRelease(connection.inputStream.bufferedReader().use { it.readText() })
            if (isVersionNewer(release.version, BuildConfig.VERSION_NAME)) {
                UpdateCheckResult.Available(release)
            } else {
                UpdateCheckResult.Current
            }
        } finally {
            connection.disconnect()
        }
    }

    suspend fun downloadAndInstall(
        activity: Activity,
        release: AppRelease,
        onProgress: (Int?) -> Unit,
    ): InstallLaunchResult {
        require(Uri.parse(release.apkUrl).scheme.equals("https", ignoreCase = true)) {
            "The release APK does not use HTTPS"
        }
        val downloadManager = appContext.getSystemService(DownloadManager::class.java)
            ?: error("Android Download Manager is unavailable")
        val safeVersion = release.version.replace(Regex("[^0-9A-Za-z._-]"), "_")
        val fileName = "UAC-SNI-Spoofer-$safeVersion-${System.currentTimeMillis()}.apk"
        val request = DownloadManager.Request(Uri.parse(release.apkUrl))
            .setTitle("UAC SNI Spoofer ${release.version}")
            .setDescription("Downloading the verified GitHub release")
            .setMimeType(APK_MIME_TYPE)
            .setAllowedOverMetered(true)
            .setAllowedOverRoaming(false)
            .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            .setDestinationInExternalFilesDir(appContext, Environment.DIRECTORY_DOWNLOADS, fileName)

        val downloadId = downloadManager.enqueue(request)
        preferences.edit().putLong(KEY_PENDING_DOWNLOAD_ID, downloadId).apply()
        withContext(Dispatchers.Main.immediate) { onProgress(0) }

        return try {
            var lastProgress: Int? = -1
            while (true) {
                val snapshot = queryDownload(downloadManager, downloadId)
                when (snapshot.status) {
                    DownloadManager.STATUS_SUCCESSFUL -> break
                    DownloadManager.STATUS_FAILED -> error("Download failed (code ${snapshot.reason})")
                    DownloadManager.STATUS_PENDING,
                    DownloadManager.STATUS_PAUSED,
                    DownloadManager.STATUS_RUNNING -> {
                        if (snapshot.progress != lastProgress) {
                            lastProgress = snapshot.progress
                            withContext(Dispatchers.Main.immediate) { onProgress(snapshot.progress) }
                        }
                    }
                    else -> error("Download stopped unexpectedly")
                }
                delay(DOWNLOAD_POLL_MS)
            }
            withContext(Dispatchers.Main.immediate) {
                openPendingInstaller(activity)
                    ?: error("The downloaded APK could not be opened")
            }
        } catch (error: Throwable) {
            preferences.edit().remove(KEY_PENDING_DOWNLOAD_ID).apply()
            throw error
        }
    }

    private fun parseRelease(jsonText: String): AppRelease {
        val root = JSONObject(jsonText)
        val tagName = root.getString("tag_name")
        val version = extractVersion(tagName)
            ?: extractVersion(root.optString("name"))
            ?: error("The latest release has no readable version")
        val assets = root.getJSONArray("assets")
        var selected: JSONObject? = null
        var selectedScore = Int.MIN_VALUE
        for (index in 0 until assets.length()) {
            val asset = assets.getJSONObject(index)
            val name = asset.optString("name")
            if (!name.endsWith(".apk", ignoreCase = true)) continue
            val lowerName = name.lowercase()
            val score = when {
                "unsigned" in lowerName -> 0
                lowerName == "uac-spoofer.apk" -> 100
                "release" in lowerName -> 80
                else -> 60
            }
            if (score > selectedScore) {
                selected = asset
                selectedScore = score
            }
        }
        val apk = selected?.takeIf { selectedScore > 0 }
            ?: error("Release $tagName does not include a signed APK")
        return AppRelease(
            version = version,
            tagName = tagName,
            title = root.optString("name").ifBlank { "UAC SNI Spoofer $version" },
            notes = root.optString("body"),
            apkName = apk.getString("name"),
            apkUrl = apk.getString("browser_download_url"),
            releaseUrl = root.getString("html_url"),
        )
    }

    private data class DownloadSnapshot(val status: Int, val reason: Int, val progress: Int?)

    private fun queryDownload(downloadManager: DownloadManager, downloadId: Long): DownloadSnapshot {
        downloadManager.query(DownloadManager.Query().setFilterById(downloadId)).use { cursor ->
            check(cursor != null && cursor.moveToFirst()) { "Downloaded file is no longer available" }
            val status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS))
            val reason = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_REASON))
            val received = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR))
            val total = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES))
            val progress = if (total > 0L) ((received * 100L) / total).toInt().coerceIn(0, 100) else null
            return DownloadSnapshot(status, reason, progress)
        }
    }

    companion object {
        internal const val REPOSITORY_URL = "https://github.com/Floxu1/UAC-SNI-Spoofer-Android"
        internal const val RELEASES_URL = "$REPOSITORY_URL/releases"
        private const val LATEST_RELEASE_API = "https://api.github.com/repos/Floxu1/UAC-SNI-Spoofer-Android/releases/latest"
        private const val APK_MIME_TYPE = "application/vnd.android.package-archive"
        private const val PREFERENCES_NAME = "app_update_state"
        private const val KEY_PENDING_DOWNLOAD_ID = "pending_download_id"
        private const val NETWORK_TIMEOUT_MS = 12_000
        private const val DOWNLOAD_POLL_MS = 450L

        fun resumePendingInstall(activity: Activity) {
            runCatching { openPendingInstaller(activity) }
        }

        private fun openPendingInstaller(activity: Activity): InstallLaunchResult? {
            val preferences = activity.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
            val downloadId = preferences.getLong(KEY_PENDING_DOWNLOAD_ID, -1L)
            if (downloadId < 0L) return null
            val downloadManager = activity.getSystemService(DownloadManager::class.java) ?: return null
            val status = downloadManager.query(DownloadManager.Query().setFilterById(downloadId)).use { cursor ->
                if (cursor == null || !cursor.moveToFirst()) return null
                cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS))
            }
            if (status != DownloadManager.STATUS_SUCCESSFUL) return null

            if (
                Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
                !activity.packageManager.canRequestPackageInstalls()
            ) {
                activity.startActivity(
                    Intent(
                        Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES,
                        Uri.parse("package:${activity.packageName}"),
                    ),
                )
                return InstallLaunchResult.PERMISSION_REQUESTED
            }

            val apkUri = downloadManager.getUriForDownloadedFile(downloadId)
                ?: error("Downloaded APK URI is unavailable")
            val installIntent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(apkUri, APK_MIME_TYPE)
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
            }
            activity.startActivity(installIntent)
            preferences.edit().remove(KEY_PENDING_DOWNLOAD_ID).apply()
            return InstallLaunchResult.INSTALLER_OPENED
        }
    }
}

internal fun isVersionNewer(candidate: String, current: String): Boolean {
    val candidateParts = versionParts(candidate)
    val currentParts = versionParts(current)
    if (candidateParts.isEmpty() || currentParts.isEmpty()) return false
    for (index in 0 until max(candidateParts.size, currentParts.size)) {
        val candidatePart = candidateParts.getOrElse(index) { 0 }
        val currentPart = currentParts.getOrElse(index) { 0 }
        if (candidatePart != currentPart) return candidatePart > currentPart
    }
    return false
}

private fun versionParts(raw: String): List<Int> =
    extractVersion(raw)?.split('.')?.mapNotNull(String::toIntOrNull).orEmpty()

private fun extractVersion(raw: String): String? =
    Regex("\\d+(?:\\.\\d+)+").find(raw)?.value
